### PR TITLE
Remove Referenceable

### PR DESCRIPTION
**Pull request by pepibumur**
Resolves https://github.com/xcodeswift/xcproj/issues/31

### Short description 📝
Objects references were duplicated. They were defined in:
- The dictionary key that references the object.
- The reference property of the object.

By having them in two places it was possible to have an object in a dictionary under a key that didn't match the object reference. 

The goal of this PR is to remove such duplication aiming for a safer API, and make it easier for developers to use `xcproj` to generate deterministic references when creating new objects.

### Solution 📦
The `Referenceable` protocol has been removed. `PBXObject`s don't have a reference property anymore. The reference is the key in the dictionary that contains the object. The creation of the pair reference-object is done:
- When the project is parsed.
- When the developer adds a new object with a given reference. 

### Implementation 👩‍💻👨‍💻
- [x] Remove `Referenceable` protocol and update the classes accordingly
- [x] Remove deprecated methods *(read below)*.
- [x] Fix broken tests.

> ⚠️ This update will require a major bump in the version since it deletes and updates classes and methods from the public interface. 

### GIF
gif

---
This change is [Reviewable](https://reviewable.io/reviews/xcodeswift/xcproj/185)

### CODE REVIEW

File: Sources/xcproj/PBXGroup.swift [hunk 1]
```diff
@@ -39,13 +37,12 @@ final public class PBXGroup: PBXFileElement {
         self.usesTabs = usesTabs
         self.indentWidth = indentWidth
         self.tabWidth = tabWidth
-        super.init(reference: reference, sourceTree: sourceTree, path: path, name: name)
+        super.init(sourceTree: sourceTree, path: path, name: name)
     }
 
     public static func == (lhs: PBXGroup,
```
Comment (toshi0383) [inline Sources/xcproj/PBXGroup.swift]: Just wondering: Is there any reason why your not using AutoEquatable from Sourcery?

File: Sources/xcproj/PBXObject.swift [hunk 2]
```diff
@@ -1,27 +1,19 @@
...
-    public let reference: String
-
-    init(reference: String) {
-        self.reference = reference
-    }
+public class PBXObject: Decodable {
 
+    // MARK: - Init
+    
+    init() {}
```
Comment (yonaskolb) [inline Sources/xcproj/PBXObject.swift]: This class should not actually be initiable as it's abstract, so we can remove this

File: Package.resolved [hunk 3]
```diff
@@ -10,6 +10,15 @@
           "version": "4.1.0"
         }
       },
+      {
+        "package": "BrightFutures",
```
Comment (yonaskolb) [inline Package.resolved]: This is from a different PR isn't it?

File: Sources/xcproj/PBXProj+Helpers.swift [hunk 4]
```diff
@@ -110,6 +110,17 @@ extension PBXProj {
             return type?.rawValue
         }
     }
+    
+    /// Returns the object with the given configuration list (project or target)
+    ///
+    /// - Parameter reference: configuration list reference.
+    /// - Returns: target or project with the given configuration list.
+    func objectWithConfigurationList(reference: String) -> PBXObject? {
```
Comment (yonaskolb) [inline Sources/xcproj/PBXProj+Helpers.swift]: Should this be under objects, as #84 moves all these helpers

File: Sources/xcproj/PBXProj.swift [hunk 5]
```diff
@@ -73,33 +74,65 @@ final public class PBXProj: Decodable {
...
         }
+        
+        /// Generates a deterministic reference from an object type and identifier.
+        /// It ensures that the generated reference doesn't collide with any existing one.
+        ///
+        /// - Parameters:
+        ///   - element: object type.
+        ///   - id: object identifier (e.g. path or name)
+        /// - Returns: reference.
+        public func generateReference<T: PBXObject>(_ element: T.Type, _ id: String) -> String {
```
Comment (yonaskolb) [inline Sources/xcproj/PBXProj.swift]: Will the old generation function get deleted?

File: Sources/xcproj/PBXGroup.swift (hunk 1 above)
Comment (pepibumur) [reply Sources/xcproj/PBXGroup.swift]: Not really. I actually think it's a good idea to automate the generation of init and the conformance of Equatable. Creating an issue 😛

File: Sources/xcproj/PBXObject.swift (hunk 2 above)
Comment (pepibumur) [reply Sources/xcproj/PBXObject.swift]: Good point, removing.

File: Package.resolved (hunk 3 above)
Comment (pepibumur) [reply Package.resolved]: Yeah, reverting.

File: Sources/xcproj/PBXProj.swift (hunk 5 above)
Comment (pepibumur) [reply Sources/xcproj/PBXProj.swift]: Yes it will, I'll remove it.

File: Sources/xcproj/PBXProj+Helpers.swift (hunk 4 above)
Comment (pepibumur) [reply Sources/xcproj/PBXProj+Helpers.swift]: Do you mean [this PR](https://github.com/xcodeswift/xcproj/pull/184/files)? Yeah it should be there. My idea was to merge that PR first and then update this one moving the helper under objects.

File: Sources/xcproj/PBXGroup.swift (hunk 1 above)
Comment (pepibumur) [reply Sources/xcproj/PBXGroup.swift]: https://github.com/xcodeswift/xcproj/issues/186

File: Sources/xcproj/PBXObject.swift (hunk 2 above)
Comment (yonaskolb) [reply Sources/xcproj/PBXObject.swift]: Ah ok. I guess it’s not public anyway so not a big issue

File: Sources/xcproj/PBXProj+Helpers.swift (hunk 4 above)
Comment (yonaskolb) [reply Sources/xcproj/PBXProj+Helpers.swift]: Yeah, whoops, meant that one

Comment (pepibumur) [reply Sources/xcproj/PBXProj+Helpers.swift]: Nice, I'll merge that one first and move this function before merging 👍

File: Sources/xcproj/PBXProj.swift (hunk 5 above)
Comment (pepibumur) [reply Sources/xcproj/PBXProj.swift]: Removed

File: Sources/xcproj/PBXProj.swift [hunk 6]
```diff
@@ -111,14 +111,14 @@ final public class PBXProj: Decodable {
         /// It ensures that the generated reference doesn't collide with any existing one.
         ///
         /// - Parameters:
-        ///   - element: object type.
+        ///   - object: the object to generate the reference for.
         ///   - id: object identifier (e.g. path or name)
         /// - Returns: reference.
-        public func generateReference<T: PBXObject>(_ element: T.Type, _ id: String) -> String {
+        public func generateReference<T: PBXObject>(_ object: T, _ id: String) -> String {
```
Comment (yonaskolb) [inline Sources/xcproj/PBXProj.swift]: No need for this function to be generic anymore. T can just be a PBXObject

Comment (pepibumur) [reply Sources/xcproj/PBXProj.swift]: Lol @yonaskolb I just updated it before I saw your comment. It's not generic anymore.

**Pull request merged**